crates/dslab-iaas/src/experiments.rs

```diff
@@ -0,0 +1,160 @@
+//! Tools for launching experiments with multiple environment configurations
+
+use std::cell::RefCell;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use dyn_clone::{clone_trait_object, DynClone};
+use log::Level;
+use serde::{Deserialize, Serialize};
+use sugars::{rc, refcell};
+
+use dslab_core::Simulation;
+
+use crate::core::config::exp_config::ExperimentConfig;
+use crate::core::config::sim_config::SimulationConfig;
+use crate::core::logger::{FileLogger, Logger, StdoutLogger};
+use crate::simulation::CloudSimulation;
+
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum Dict {
+    Value(String),
+    Dict(HashMap<String, String>),
+}
+
+/// Callbacks on simulation events
+pub trait SimulationCallbacks: DynClone + Send {
+    fn on_simulation_start(&mut self, _sim: Rc<RefCell<CloudSimulation>>) {
+        // custom callback
+    }
+
+    // if returns false then the simulation is stopped
+    fn on_step(&mut self, _sim: Rc<RefCell<CloudSimulation>>) -> bool {
+        // custom callback
+        true
+    }
+
+    /// returns experiment test case metrics
+    fn on_simulation_finish(&mut self, _sim: Rc<RefCell<CloudSimulation>>) -> HashMap<String, String> {
+        // custom callback
+        HashMap::new()
+    }
+}
+
+clone_trait_object!(SimulationCallbacks);
+
+fn process_test_case(
+    current_config: SimulationConfig,
+    callbacks: &mut Box<dyn SimulationCallbacks>,
+    log_file: Option<String>,
+    log_level: Option<Level>,
+) -> HashMap<String, String> {
+    let mut logger: Box<dyn Logger> = Box::new(StdoutLogger::with_level(log_level.unwrap_or(Level::Info)));
+    if log_file.is_some() {
+        logger = Box::new(FileLogger::with_level(log_level.unwrap_or(Level::Info)));
+    }
+
+    let sim = Simulation::new(123);
+    let cloud_sim = rc!(refcell!(CloudSimulation::with_logger(
+        sim,
+        current_config.clone(),
+        logger
+    )));
+    callbacks.on_simulation_start(cloud_sim.clone());
+
+    while cloud_sim.borrow_mut().current_time() < current_config.simulation_length {
+        cloud_sim.borrow_mut().steps(1);
+        let proceed = callbacks.on_step(cloud_sim.clone());
+        if !proceed {
+            break;
+        }
+    }
+
+    let result = callbacks.on_simulation_finish(cloud_sim.clone());
+    if log_file.is_some() {
+        let save_result = cloud_sim.borrow_mut().save_trace_to_file(&log_file.clone().unwrap());
```

Comment:
Мутабельный borrow не требуется

```suggestion
        let save_result = cloud_sim.borrow().save_trace_to_file(&log_file.clone().unwrap());
```
и внутри `save_trace_to_file` тоже заменить на `borrow`